File: packages/android/src/main/java/com/kozmos/components/Spinner/Spinner.kt
package com.kozmos.components.spinner

import android.provider.Settings
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

/**
 * Layout of the spinner, mirroring the React `Spinner.size`.
 *
 * The `Spinner/size` scale (small, medium, large, xlarge): 16, 24, 32 and 48,
 * as React's classes and the Figma variables draw them. The natives have no
 * generated token for these — the four live as Figma component variables — so
 * the scale is written here and on SwiftUI until they reach `packages/tokens`.
 *
 * The doc sits above the declaration, not inside the body: the variant-parity
 * check splits an enum body on commas to read its entries, and a comment
 * holding one took Spinner's whole axis with it.
 */
enum class KozmosSpinnerSize(val value: String, val dimension: Dp) {
    Sm("sm", 16.dp),
    Md("md", 24.dp),
    Lg("lg", 32.dp),
    Xl("xl", 48.dp)
}

/**
 * Indeterminate loading: the system's arc, turning.
 *
 * Before 2026-09-22 this was material3's `CircularProgressIndicator` at its own
 * default size, while React drew lucide's arc, iOS drew a `ProgressView` tinted
 * a hard-coded blue and Figma drew an ellipse with a dash pattern. Four
 * drawings, no two alike — and material3 1.1.2's indeterminate indicator is the
 * one Paparazzi cannot render against animation-core 1.6.0, so no golden has
 * ever shown one.
 *
 * This is the one drawing: three quarters of a circle from the top, round caps,
 * stroke 2 in the icons' own 24 box so the weight scales with the mark. It
 * rests when the system's animations are off; the description still says what
 * is happening.
 */
@Composable
fun KozmosSpinner(
    modifier: Modifier = Modifier,
    size: KozmosSpinnerSize = KozmosSpinnerSize.Md,
    label: String = "Loading",
    // The theme's own text colour, as React's `currentColor` and SwiftUI's
    // foreground style resolve to. It defaulted to theme/500 whatever it sat
    // on, which was the last place the three platforms disagreed once the arc
    // was one drawing. `LocalContentColor` is the closer-looking analogue and
    // is wrong here: outside a material `Surface` it stays its own default,
    // black, which drew an invisible spinner on the dark surface — the first
    // dark golden was a black rectangle. A control that sets a content colour
    // — a Button, an IconButton — passes it in.
    color: Color = KozmosThemeTokens.primitivesColorsForeground0
) {
    // One turn a second, linear — the same turn React's `kozmos-spin` takes.
    val transition = rememberInfiniteTransition(label = "kozmos-spinner")
    val angle by transition.animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 1000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "kozmos-spinner-angle"
    )
    val context = LocalContext.current
    val animationsOn = remember(context) {
        Settings.Global.getFloat(
            context.contentResolver,
            Settings.Global.ANIMATOR_DURATION_SCALE,
            1f
        ) > 0f
    }
    val strokeWidth = KozmosDimensions.primitivesIconStrokeMd * (size.dimension / 24.dp)
    Canvas(
        modifier = modifier
            .size(size.dimension)
            .semantics { contentDescription = label }
    ) {
        val stroke = strokeWidth.toPx()
        // Inset by half the stroke: an arc is centred on its path, so a circle
        // drawn on the edge loses half its width off the side.
        val inset = stroke / 2f
        drawArc(
            color = color,
            // From the top. Compose measures from 3 o'clock, so -90 puts the
            // gap at the end of three quarters rather than across the top.
            startAngle = -90f + if (animationsOn) angle else 0f,
            sweepAngle = 270f,
            useCenter = false,
            topLeft = Offset(inset, inset),
            size = Size(this.size.width - stroke, this.size.height - stroke),
            style = Stroke(width = stroke, cap = StrokeCap.Round)
        )
    }
}

File: packages/ios/Sources/Components/Spinner/Spinner.swift
import SwiftUI

/// Layout of the spinner, mirroring the React `Spinner.size`.
public enum KozmosSpinnerSize: String, CaseIterable, Sendable {
    case sm
    case md
    case lg
    case xl

    /// The `Spinner/size` scale (small, medium, large, xlarge): 16, 24, 32 and 48, as React's classes and
    /// the Figma variables draw them. The natives have no generated token for
    /// these — the four live as Figma component variables — so the scale is
    /// written here and on Compose until they reach `packages/tokens`.
    var points: CGFloat {
        switch self {
        case .sm: return 16
        case .md: return 24
        case .lg: return 32
        case .xl: return 48
        }
    }

    /// Stroke 2 in the icons' own 24 box, so the weight scales with the mark
    /// exactly as every Kozmos icon's does: `KozmosDimensions
    /// .primitivesIconStrokeMd` at 24, proportionally above and below.
    var lineWidth: CGFloat {
        KozmosDimensions.primitivesIconStrokeMd * points / 24
    }
}

/// Indeterminate loading: the system's arc, turning.
///
/// Before 2026-09-22 this was `ProgressView().tint(.blue)` — a hard-coded
/// SwiftUI blue that ignored the theme, at a size the caller could not set,
/// drawn as iOS's twelve-spoke wheel while React drew lucide's arc, Compose
/// drew material3's and Figma drew a dashed ellipse. Four drawings, no two
/// alike.
///
/// This is the one drawing: three quarters of a circle, round caps, taking its
/// colour from the foreground it sits in, so a spinner on a themed surface
/// follows the theme without being told. It rests when the visitor has asked
/// for less motion; the label still says what is happening.
public struct KozmosSpinner: View {
    private let size: KozmosSpinnerSize
    private let color: Color
    private let label: String

    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var angle: Double = 0

    /// One turn a second, linear — the same turn React's `kozmos-spin` takes.
    static let turnDuration: Double = 1

    /// The colour is the theme's own text colour, as React's `currentColor` and
    /// Compose's themed foreground resolve to — not SwiftUI's implicit
    /// `.primary`, which follows the system appearance rather than the theme a
    /// Kozmos surface is drawing in. A control that has a foreground of its own
    /// — a button, an icon button — passes it in.
    public init(
        size: KozmosSpinnerSize = .md,
        color: Color = KozmosColors.primitivesColorsForeground0,
        label: String = "Loading"
    ) {
        self.size = size
        self.color = color
        self.label = label
    }

    public var body: some View {
        Circle()
            // From the top, three quarters of the way round. The quarter that
            // is missing is what reads as motion; a full ring turning shows
            // nothing at all.
            .trim(from: 0, to: 0.75)
            .stroke(color, style: StrokeStyle(lineWidth: size.lineWidth, lineCap: .round))
            .rotationEffect(.degrees(-90 + angle))
            .frame(width: size.points, height: size.points)
            .onAppear {
                guard !reduceMotion else { return }
                withAnimation(.linear(duration: Self.turnDuration).repeatForever(autoreverses: false)) {
                    angle = 360
                }
            }
            .accessibilityElement()
            .accessibilityLabel(label)
            .accessibilityAddTraits(.updatesFrequently)
    }
}
